numpyro/infer/elbo.py

```diff
@@ -311,53 +312,115 @@ class RenyiELBO(ELBO):
             model_seed, guide_seed = random.split(rng_key)
             seeded_model = seed(model, model_seed)
             seeded_guide = seed(guide, guide_seed)
-            guide_log_density, guide_trace = log_density(
-                seeded_guide, args, kwargs, param_map
-            )
-            # NB: we only want to substitute params not available in guide_trace
-            model_param_map = {
-                k: v for k, v in param_map.items() if k not in guide_trace
-            }
-            seeded_model = replay(seeded_model, guide_trace)
-            model_log_density, model_trace = log_density(
-                seeded_model, args, kwargs, model_param_map
+            model_trace, guide_trace = get_importance_trace(
+                seeded_model, seeded_guide, args, kwargs, param_map
             )
             check_model_guide_match(model_trace, guide_trace)
             _validate_model(model_trace, plate_warning="loose")
 
+            site_plates = {
+                name: {frame for frame in site["cond_indep_stack"]}
+                for name, site in model_trace.items()
+                if site["type"] == "sample"
+            }
+            if site_plates:
+                common_plates = set.intersection(*site_plates.values())
+            else:
+                common_plates = set()
+            common_plate_scale = 1.0
+            for frame in common_plates:
+                subsample_size = frame.size
+                size = model_trace[frame.name]["args"][0]
+                if size > subsample_size and frame.name not in indep_subsample_plates:
+                    common_plates = common_plates - {frame}
+                else:
+                    common_plate_scale = common_plate_scale * size / subsample_size
+            common_plate_dims = {frame.dim for frame in common_plates}
+
+            log_densities = {}
+            for trace_type, tr in {"guide": guide_trace, "model": model_trace}.items():
+                log_densities[trace_type] = 1.0
+                for site in tr.values():
+                    if site["type"] != "sample":
+                        continue
+                    log_prob = site["log_prob"]
+                    squeeze_axes = ()
+                    for dim in range(log_prob.ndim):
+                        neg_dim = dim - log_prob.ndim
+                        if neg_dim in common_plate_dims:
+                            continue
+                        log_prob = jnp.sum(log_prob, axis=dim, keepdims=True)
+                        squeeze_axes = squeeze_axes + (dim,)
+                    log_prob = jnp.squeeze(log_prob, squeeze_axes)
+                    log_densities[trace_type] = log_densities[trace_type] + log_prob
+
             # log p(z) - log q(z)
-            elbo = model_log_density - guide_log_density
-            return elbo
+            elbo = log_densities["model"] - log_densities["guide"]
+            # Do not scale elbo at common plate dimensions.
```

Comment:
doesn't this sort of break the pyro dimension? what scale is this exactly?

Reply:
I added a better test to illustrate this. Assume "N" is a plate with size N and subsample size n, the scale is N / n. We store the scaled log probabilities in the trace. Here we remove such scaling operator because we don't want to perform Monte Carlo estimation at this dimension before computing renyi_elbo. We will perform MC estimation after computing renyi_elbo for each item of this plate.